File: src/main/java/com/leet/medium/CloneGraph.go
package main

/**
 * Definition for a Node.
 */

type NAryNode struct {
	Val       int
	Neighbors []*NAryNode
}

func cloneGraphUtil(node *NAryNode, mapping []*NAryNode) *NAryNode {
	if node == nil {
		return nil
	}

	if cloned := mapping[node.Val]; cloned != nil {
		return cloned
	}

	clone := &NAryNode{Val: node.Val, Neighbors: make([]*NAryNode, len(node.Neighbors))}
	mapping[node.Val] = clone

	for i, neighbor := range node.Neighbors {
		clone.Neighbors[i] = cloneGraphUtil(neighbor, mapping)
	}

	return clone
}

func cloneGraph(node *NAryNode) *NAryNode {
	mapping := make([]*NAryNode, 100+1)
	ans := cloneGraphUtil(node, mapping)
	return ans
}

File: src/main/java/com/leet/medium/CloneGraph.ts
/**
 * Definition for Node.
 */
class Node {
  val: number;
  neighbors: Node[];
  constructor(val?: number, neighbors?: Node[]) {
    this.val = val === undefined ? 0 : val;
    this.neighbors = neighbors === undefined ? [] : neighbors;
  }
}
function cloneNode(node: Node, clones: Map<Node, Node>): Node {
  if (clones.has(node)) {
    return clones.get(node) as Node;
  }
  const clone = new Node(node.val);
  clones.set(node, clone);
  clone.neighbors = node.neighbors.map((neighbor) =>
    cloneNode(neighbor, clones)
  ) as Node[];
  return clone;
}

function cloneGraph(node: Node | null): Node | null {
  if (!node) return node;
  return cloneNode(node, new Map());
}
